`evaluation panicked: FormatElement size exceeds 24 bytes` when building oxlint version 1.29.0 on 32-bit x86 GNU/Linux

<!--
!!IMPORTANT!! If you are reporting a bug for Oxlint or the Oxc VSCode extension,
please use the "Linter bug report" template instead. You can find it here:
https://github.com/oxc-project/oxc/issues/new?labels=C-bug,A-linter&projects=&template=linter_bug_report.yaml&title=linter:+
-->

Hi, I wanted to use oxlint version 1.29.0 on 32-bit x86 GNU/Linux, but this error occurred:

```
error[E0080]: evaluation panicked: `FormatElement` size exceeds 24 bytes, expected 24 bytes in 32-bit platforms
  --> crates/oxc_formatter/src/formatter/format_element/mod.rs:29:9
   |
29 | /         assert!(
30 | |             size_of::<FormatElement>() == 24,
31 | |             "`FormatElement` size exceeds 24 bytes, expected...
32 | |         );
   | |_________^ evaluation of `formatter::format_element::_` failed here
```

This error does not occur with oxlint version 1.27.0.

I would guess this error might be related to this PR:

- https://github.com/oxc-project/oxc/pull/15367

oxlint version: 1.29.0 and git main

Operating system:

```
$ cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 13 (trixie)"
NAME="Debian GNU/Linux"
VERSION_ID="13"
VERSION="13 (trixie)"
VERSION_CODENAME=trixie
DEBIAN_VERSION_FULL=13.1
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
$ uname -a
Linux debian 6.1.0-32-686-pae #1 SMP PREEMPT_DYNAMIC Debian 6.1.129-1 (2025-03-06) i686 GNU/Linux
$ cargo --version
cargo 1.91.1 (ea2d97820 2025-10-10)
$ rustc --version
rustc 1.91.1 (ed61e7d7e 2025-11-07)
$ 
```

I have double-checked again just now using the main branch at commit 2bfdd26ca37383068c8052cdfc2884f0f9983408 by deleting the repository and redownloading it, and here is the complete log:

<details>
<summary>log</summary>

```
tacokoneko@debian:~$ git clone https://github.com/oxc-project/oxc
Cloning into 'oxc'...
remote: Enumerating objects: 200989, done.
remote: Counting objects: 100% (1364/1364), done.
remote: Compressing objects: 100% (755/755), done.
remote: Total 200989 (delta 947), reused 612 (delta 609), pack-reused 199625 (from 2)
Receiving objects: 100% (200989/200989), 491.59 MiB | 28.76 MiB/s, done.
Resolving deltas: 100% (147572/147572), done.
tacokoneko@debian:~$ cd oxc
tacokoneko@debian:~/oxc$ cargo build
    Updating crates.io index
  Downloaded oxc-schemars v0.8.27
  Downloaded 1 crate (58.7KiB) in 0.65s
   Compiling proc-macro2 v1.0.103
   Compiling quote v1.0.42
   Compiling unicode-ident v1.0.22
   Compiling cfg-if v1.0.4
   Compiling memchr v2.7.6
   Compiling serde_core v1.0.228
   Compiling allocator-api2 v0.2.21
   Compiling itoa v1.0.15
   Compiling crossbeam-utils v0.8.21
   Compiling equivalent v1.0.2
   Compiling ryu v1.0.20
   Compiling foldhash v0.2.0
   Compiling serde v1.0.228
   Compiling rayon-core v1.13.0
   Compiling either v1.15.0
   Compiling hashbrown v0.16.0
   Compiling siphasher v1.0.1
   Compiling unicode-segmentation v1.12.0
   Compiling smallvec v1.15.1
   Compiling serde_json v1.0.145
   Compiling phf_shared v0.13.1
   Compiling rustc-hash v2.1.1
   Compiling thiserror v2.0.17
   Compiling owo-colors v4.2.3
   Compiling crossbeam-epoch v0.9.18
   Compiling regex-syntax v0.8.8
   Compiling rustversion v1.0.22
   Compiling syn v2.0.110
   Compiling crossbeam-deque v0.8.6
   Compiling unicode-width v0.2.2
   Compiling aho-corasick v1.1.4
   Compiling smawk v0.3.2
   Compiling unicode-linebreak v0.1.5
   Compiling castaway v0.2.4
   Compiling textwrap v0.16.2
   Compiling rayon v1.11.0
   Compiling phf v0.13.1
   Compiling bumpalo v3.19.0
   Compiling static_assertions v1.1.0
   Compiling str_indices v0.4.4
   Compiling autocfg v1.5.0
   Compiling ropey v1.6.1
   Compiling num-traits v0.2.19
   Compiling fastrand v2.3.0
   Compiling bitflags v2.10.0
   Compiling regex-automata v0.4.13
   Compiling oxc_data_structures v0.98.0 (/home/tacokoneko/oxc/crates/oxc_data_structures)
   Compiling percent-encoding v2.3.2
   Compiling dragonbox_ecma v0.0.5
   Compiling oxc-schemars v0.8.27
   Compiling oxc_estree v0.98.0 (/home/tacokoneko/oxc/crates/oxc_estree)
   Compiling phf_generator v0.13.1
   Compiling dyn-clone v1.0.20
   Compiling cow-utils v0.1.3
   Compiling nonmax v0.5.5
   Compiling unicode-id-start v1.4.0
   Compiling napi-build v2.3.1
   Compiling libc v0.2.177
   Compiling foldhash v0.1.5
   Compiling once_cell v1.21.3
   Compiling hashbrown v0.15.5
   Compiling pin-project-lite v0.2.16
   Compiling itertools v0.14.0
   Compiling fixedbitset v0.5.7
   Compiling indexmap v2.12.0
   Compiling num-integer v0.1.46
   Compiling futures-core v0.3.31
   Compiling futures-sink v0.3.31
   Compiling log v0.4.28
   Compiling num-bigint v0.4.6
   Compiling self_cell v1.2.1
   Compiling futures-channel v0.3.31
   Compiling pin-utils v0.1.0
   Compiling seq-macro v0.3.6
   Compiling slab v0.4.11
   Compiling futures-task v0.3.31
   Compiling futures-io v0.3.31
   Compiling convert_case v0.9.0
   Compiling petgraph v0.8.3
   Compiling dtor-proc-macro v0.0.6
   Compiling regex v1.12.2
   Compiling dtor v0.1.1
   Compiling ctor-proc-macro v0.0.7
   Compiling serde_derive_internals v0.29.1
   Compiling ctor v0.6.1
   Compiling synstructure v0.13.2
   Compiling napi v3.5.2
   Compiling libloading v0.9.0
   Compiling semver v1.0.27
   Compiling napi-sys v3.1.1
   Compiling napi-derive-backend v3.0.2
   Compiling nohash-hasher v0.2.0
   Compiling oxc_sourcemap v6.0.1
   Compiling vsimd v0.8.0
   Compiling find-msvc-tools v0.1.4
   Compiling bytes v1.10.1
   Compiling outref v0.5.2
   Compiling shlex v1.3.0
   Compiling httparse v1.10.1
   Compiling base64-simd v0.8.0
   Compiling cc v1.2.45
   Compiling stable_deref_trait v1.2.1
   Compiling json-escape-simd v3.0.1
   Compiling fnv v1.0.7
   Compiling crc32fast v1.5.0
   Compiling simd-adler32 v0.3.7
   Compiling typenum v1.19.0
   Compiling powerfmt v0.2.0
   Compiling adler2 v2.0.1
   Compiling virtue v0.0.18
   Compiling version_check v0.9.5
   Compiling deranged v0.5.5
   Compiling miniz_oxide v0.8.9
   Compiling generic-array v0.14.9
   Compiling serde_derive v1.0.228
   Compiling thiserror-impl v2.0.17
   Compiling oxc-miette-derive v2.6.0
   Compiling oxc_ast_macros v0.98.0 (/home/tacokoneko/oxc/crates/oxc_ast_macros)
   Compiling oxc_schemars_derive v0.8.26
   Compiling phf_macros v0.13.1
   Compiling futures-macro v0.3.31
   Compiling tokio-macros v2.6.0
   Compiling napi-derive v3.3.3
   Compiling oxc-miette v2.6.0
   Compiling zerofrom-derive v0.1.6
   Compiling yoke-derive v0.8.1
   Compiling zerovec-derive v0.11.2
   Compiling futures-util v0.3.31
   Compiling tokio v1.48.0
   Compiling oxc_diagnostics v0.98.0 (/home/tacokoneko/oxc/crates/oxc_diagnostics)
   Compiling displaydoc v0.2.5
   Compiling bincode_derive v2.0.1
   Compiling time-core v0.1.6
   Compiling writeable v0.6.2
   Compiling num-conv v0.1.0
   Compiling zerofrom v0.1.6
   Compiling yoke v0.8.1
   Compiling base64 v0.22.1
   Compiling unty v0.0.4
   Compiling time v0.3.44
   Compiling flate2 v1.1.5
   Compiling bincode v2.0.1
   Compiling zerovec v0.11.5
   Compiling nom v8.0.0
   Compiling oxc_traverse v0.98.0 (/home/tacokoneko/oxc/crates/oxc_traverse)
   Compiling compact_str v0.9.0
   Compiling oxc_allocator v0.98.0 (/home/tacokoneko/oxc/crates/oxc_allocator)
   Compiling oxc_index v4.1.0
   Compiling futures-executor v0.3.31
   Compiling futures v0.3.31
   Compiling tinystr v0.8.2
   Compiling ring v0.17.14
   Compiling litemap v0.8.1
   Compiling oxc_span v0.98.0 (/home/tacokoneko/oxc/crates/oxc_span)
   Compiling icu_locale_core v2.1.1
   Compiling potential_utf v0.1.4
   Compiling zerotrie v0.2.3
   Compiling libm v0.2.15
   Compiling oxc_syntax v0.98.0 (/home/tacokoneko/oxc/crates/oxc_syntax)
   Compiling oxc_regular_expression v0.98.0 (/home/tacokoneko/oxc/crates/oxc_regular_expression)
   Compiling zeroize v1.8.2
   Compiling same-file v1.0.6
   Compiling walkdir v2.5.0
   Compiling rustls-pki-types v1.13.0
   Compiling icu_provider v2.1.1
   Compiling icu_collections v2.1.1
   Compiling crypto-common v0.1.6
   Compiling oxc_ast v0.98.0 (/home/tacokoneko/oxc/crates/oxc_ast)
   Compiling oxc_cfg v0.98.0 (/home/tacokoneko/oxc/crates/oxc_cfg)
   Compiling block-buffer v0.10.4
   Compiling getrandom v0.2.16
   Compiling untrusted v0.9.0
   Compiling oxc-browserslist v2.1.3
   Compiling digest v0.10.7
   Compiling icu_properties_data v2.1.1
   Compiling icu_normalizer_data v2.1.1
   Compiling rustls v0.23.35
   Compiling project-root v0.2.2
   Compiling oxc_compat v0.98.0 (/home/tacokoneko/oxc/crates/oxc_compat)
   Compiling cpufeatures v0.2.17
   Compiling sha1 v0.10.6
   Compiling rustls-webpki v0.103.8
   Compiling http v1.3.1
   Compiling json-strip-comments v3.0.1
   Compiling subtle v2.6.1
   Compiling oxc_formatter v0.14.0 (/home/tacokoneko/oxc/crates/oxc_formatter)
   Compiling webpki-roots v1.0.4
   Compiling tracing-core v0.1.34
   Compiling similar v2.7.0
   Compiling ureq-proto v0.5.2
   Compiling utf-8 v0.7.6
   Compiling icu_normalizer v2.1.1
   Compiling icu_properties v2.1.1
   Compiling console v0.16.1
   Compiling utf8_iter v1.0.4
   Compiling rustix v1.1.2
   Compiling icu_locale_data v2.1.1
   Compiling natord v1.0.9
   Compiling tracing-attributes v0.1.30
   Compiling float-cmp v0.10.0
   Compiling halfbrown v0.4.0
   Compiling icu_segmenter_data v2.1.1
   Compiling linux-raw-sys v0.11.0
   Compiling value-trait v0.12.1
   Compiling lazy-regex-proc_macros v3.4.2
   Compiling idna_adapter v1.2.1
   Compiling idna v1.1.0
   Compiling seize v0.5.1
   Compiling form_urlencoded v1.2.2
   Compiling bstr v1.12.1
   Compiling lazy-regex v3.4.2
   Compiling constcat v0.6.1
   Compiling simdutf8 v0.1.5
   Compiling url v2.5.7
   Compiling tracing v0.1.41
   Compiling globset v0.4.18
   Compiling simd-json v0.17.0
   Compiling papaya v0.2.3
   Compiling ureq v3.1.4
   Compiling icu_locale v2.1.1
   Compiling core_maths v0.1.1
   Compiling parking_lot_core v0.9.12
   Compiling arrayvec v0.7.6
   Compiling oxc_resolver v11.13.2
   Compiling fast-glob v1.0.0
   Compiling oxc_tasks_common v0.0.0 (/home/tacokoneko/oxc/tasks/common)
   Compiling ignore v0.4.25
   Compiling javascript-globals v1.3.4
   Compiling icu_segmenter v2.1.1
   Compiling oxc_macros v0.0.0 (/home/tacokoneko/oxc/crates/oxc_macros)
   Compiling oxc_ecmascript v0.98.0 (/home/tacokoneko/oxc/crates/oxc_ecmascript)
   Compiling oxc_ast_visit v0.98.0 (/home/tacokoneko/oxc/crates/oxc_ast_visit)
   Compiling oxc_parser v0.98.0 (/home/tacokoneko/oxc/crates/oxc_parser)
   Compiling rust-lapper v1.2.0
   Compiling scopeguard v1.2.0
   Compiling language-tags v0.3.2
   Compiling is_ci v1.2.0
   Compiling bitflags v1.3.2
   Compiling fluent-uri v0.1.4
   Compiling supports-color v3.0.2
   Compiling lock_api v0.4.14
   Compiling bpaf_derive v0.5.17
   Compiling serde_repr v0.1.20
   Compiling tower-service v0.3.3
   Compiling pico-args v0.5.0
   Compiling ident_case v1.0.1
   Compiling hashbrown v0.14.5
   Compiling strsim v0.11.1
   Compiling oxc_semantic v0.98.0 (/home/tacokoneko/oxc/crates/oxc_semantic)
   Compiling sync_wrapper v1.0.2
   Compiling tower-layer v0.3.3
   Compiling lsp-types v0.97.0
   Compiling tower v0.5.2
   Compiling darling_core v0.20.11
   Compiling dashmap v6.1.0
   Compiling oxc_isolated_declarations v0.98.0 (/home/tacokoneko/oxc/crates/oxc_isolated_declarations)
   Compiling tokio-util v0.7.17
   Compiling bpaf v0.9.20
   Compiling env_filter v0.1.4
   Compiling oxc_codegen v0.98.0 (/home/tacokoneko/oxc/crates/oxc_codegen)
   Compiling oxc_mangler v0.98.0 (/home/tacokoneko/oxc/crates/oxc_mangler)
   Compiling oxc_linter v1.29.0 (/home/tacokoneko/oxc/crates/oxc_linter)
error[E0080]: evaluation panicked: `FormatElement` size exceeds 24 bytes, expected 24 bytes in 32-bit platforms
  --> crates/oxc_formatter/src/formatter/format_element/mod.rs:29:9
   |
29 | /         assert!(
30 | |             size_of::<FormatElement>() == 24,
31 | |             "`FormatElement` size exceeds 24 bytes, expected 24 bytes in 32-bit platforms"
32 | |         );
   | |_________^ evaluation of `formatter::format_element::_` failed here

   Compiling jiff v0.2.15
   Compiling darling_macro v0.20.11
   Compiling lazy_static v1.5.0
   Compiling ucd-trie v0.1.7
   Compiling sharded-slab v0.1.7
   Compiling pest v2.8.3
   Compiling darling v0.20.11
   Compiling tracing-log v0.2.0
   Compiling thread_local v1.1.9
   Compiling getrandom v0.3.4
   Compiling nu-ansi-term v0.50.3
   Compiling zerocopy v0.8.27
   Compiling tracing-subscriber v0.3.20
   Compiling derive_builder_core v0.20.2
   Compiling pest_meta v2.8.3
   Compiling oxc_transformer v0.98.0 (/home/tacokoneko/oxc/crates/oxc_transformer)
   Compiling oxc_minifier v0.98.0 (/home/tacokoneko/oxc/crates/oxc_minifier)
   Compiling cmake v0.1.54
   Compiling zerocopy-derive v0.8.27
   Compiling env_logger v0.11.8
   Compiling oxc_napi v0.98.0 (/home/tacokoneko/oxc/crates/oxc_napi)
   Compiling libmimalloc-sys2 v0.1.51
   Compiling pest_generator v2.8.3
For more information about this error, try `rustc --explain E0080`.
error: could not compile `oxc_formatter` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
tacokoneko@debian:~/oxc$ 
```

</details>

--- a/crates/oxc_formatter/src/formatter/format_element/mod.rs
+++ b/crates/oxc_formatter/src/formatter/format_element/mod.rs
@@ -25,7 +25,11 @@ const _: () = {
             size_of::<FormatElement>() == 40,
             "`FormatElement` size exceeds 40 bytes, expected 40 bytes in 64-bit platforms"
         );
-    } else {
+    } else if cfg!(target_family = "wasm") || align_of::<u64>() == 8 {
+        // Some 32-bit platforms have 8-byte alignment for `u64` and `f64`, while others have 4-byte alignment.
+        //
+        // Skip these assertions on 32-bit platforms where `u64` / `f64` have 4-byte alignment, because
+        // some layout calculations may be incorrect. https://github.com/oxc-project/oxc/pull/13716
         assert!(
             size_of::<FormatElement>() == 24,
             "`FormatElement` size exceeds 24 bytes, expected 24 bytes in 32-bit platforms"
@@ -40,7 +44,11 @@ const _: () = {
             size_of::<FormatElement>() == 24,
             "`FormatElement` size exceeds 24 bytes, expected 24 bytes in 64-bit platforms"
         );
-    } else {
+    } else if cfg!(target_family = "wasm") || align_of::<u64>() == 8 {
+        // Some 32-bit platforms have 8-byte alignment for `u64` and `f64`, while others have 4-byte alignment.
+        //
+        // Skip these assertions on 32-bit platforms where `u64` / `f64` have 4-byte alignment, because
+        // some layout calculations may be incorrect. https://github.com/oxc-project/oxc/pull/13716
         assert!(
             size_of::<FormatElement>() == 16,
             "`FormatElement` size exceeds 16 bytes, expected 16 bytes in 32-bit platforms"

